# <center>Đồ án 3: `Linear Regression`</center>

- ##### Name: `Lý Nhật Hào`
- ##### MSSV: `21127041`
- ##### Class: `21CLC04`

Trong file.ipynb trên sẽ bao gồm:
- `source code` giải quyết yêu cầu đồ án 3 cũng như mục tiêu của đồ án đó là:
    - Tìm hiểu các yếu tố quyết định mức lương và việc làm của các kỹ sư ngay sau khi tốt nghiệp. Các yếu tố như điểm số ở các cấp/trường đại học, kỹ năng của ứng viên, sự liên kết giữa trường đại học và các khu công nghiệp/công ty công nghệ, bằng cấp của sinh viên và điều kiện thị trường cho các ngành công nghiệp cụ thể sẽ ảnh hưởng đến điều này.

    - Bộ dữ liệu được sử dụng trong đồ án này thu thập tại Ấn Độ, nơi có hơn 6000 cơ sở đào tạo kỹ thuật công nghệ với khoảng 2,9 triệu sinh viên đang học tập. Mỗi năm, trung bình có 1,5 triệu sinh viên tốt nghiệp chuyên ngành Công nghệ/Kỹ thuật, tuy nhiên do thiếu kỹ năng cần thiết, ít hơn 20% trong số họ có việc làm phù hợp với chuyên môn của mình. Bộ dữ liệu này không chỉ giúp xây dựng công cụ dự đoán mức lương mà còn cung cấp thông tin về các yếu tố ảnh hưởng đến mức lương và chức danh công việc trên thị trường lao động. Sinh viên sẽ được khám phá những thông tin này trong phạm vi đồ án.
- `Chú thích` ý nghĩa cơ bản của những hàm được cài đặt cũng như một vài biến quan trọng được cài đặt trong file code.

##### <center> <font style="color:red"> ** Những chú thích, kết quả, phân tích, nhận xét và bàn luận chi tiết của đồ án trên sẽ được triển khai sâu hơn trong file 21127041.PDF **</font><center>

# Import Thư Viện
|STT|THƯ VIỆN                | CÔNG DỤNG                   |HÀM DÙNG TRONG THƯ VIỆN
|---|----------|-------------------------------------------|----|
|1|pandas | Để xử lý nhận-trả dữ liệu theo dạng data frame|pandas.DataFrame.sample(frac = 1) ; pandas.read_csv(name of file) ; pandas.iloc[] ; pandas.copy(deep = True) ; pandas.DataFrame(data) ; pandas.corr():|
|2|numpy | Thư viện quen thuộc dùng cho ma trận, array,… trong đồ án trên là để append|numpy.mean() ; numpy.absolute()|
|3|sklearn|Để có thể sử dụng thuật toán linear regression, k_fold cross validation và để tính MAE|sklearn.linear_model.LinearRegression() ; LinearRegression.fit(X, y) ; LinearRegression.predict(X_test) ; LinearRegression.coef_ ; sklearn.model_selection ; sklearn.model_selection.cross_val_score ; sklearn.metrics.mean_absolute_error|




In [253]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_absolute_error

# Đọc dữ liệu
- Đọc dữ liệu bằng pandas
- Lấy các đặc trưng X và giá trị mục tiêu y cho các tập huấn luyện (train) và kiểm tra (test)


In [254]:
# Sinh viên có thể sử dụng các khác nếu cần

train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

X_train = train.iloc[:, :11]   
y_train = train.iloc[:, -1]     

X_test = test.iloc[:, :11]      
y_test = test.iloc[:, -1]       



# Cài đặt hàm

##### Hàm bên dưới dùng để tính MAE 

In [255]:
def calculate_mae(test, y_test_pred): # TÍNH MAE
    mae = mean_absolute_error(test.iloc[:, -1], y_test_pred)
    return mae

# Yêu cầu `1A`: Sử dụng toàn bộ 11 đặc trưng đầu tiên `Gender`, `'10percentage'`, `'12percentage'`, `'CollegeTier'`, `'Degree'`, `'collegeGPA'`, `'CollegeCityTier`, `English` , `Logical`, `Quant`, `Domain` (2 điểm) 

MỘT VÀI BƯỚC CƠ BẢN ĐỂ GIẢI QUYẾT CÂU `1A` NHƯ SAU:
- Đọc dữ liệu huấn luyện từ tệp `'train.csv'`và dữ liệu kiểm tra từ tệp `'test.csv' `bằng cách sử dụng hàm pd.read_csv(). Dữ liệu huấn luyện được lưu trong biến `training_1`, và dữ liệu kiểm tra được lưu trong biến `testing_1`.
- Tạo một mô hình hồi quy tuyến tính bằng cách sử dụng lớp `LinearRegression() `từ thư viện scikit-learn.
- Sử dụng mô hình đã được huấn luyện để dự đoán kết quả trên dữ liệu kiểm tra. Giá trị dự đoán được lưu trong biến `y_test_pred`.
- Tạo một DataFrame có tên prediction để so sánh kết quả dự đoán với giá trị thực tế trên dữ liệu kiểm tra. DataFrame `prediction `được in ra màn hình bằng hàm print(), số làm tròn đến 3 chữ số thập phân bằng phương thức round(3).
- Tương tự với `coefficent`

In [256]:
# Cài đặt các hàm cần thiết ở đây
# Phần code cho yêu cầu 1a

#đọc 2 cái file
testing_1 = pd.read_csv('test.csv')
training_1 = pd.read_csv('train.csv')


#dùng linearregression và lấy 11 tính cách đầu của bộ dữ liệu
lr1 = LinearRegression(fit_intercept=True).fit(training_1.iloc[:, :11], training_1.iloc[:, -1])

y_test_pred = lr1.predict(testing_1.iloc[:, :11])
predic = pd.DataFrame({'data from file': testing_1.iloc[:, -1], 'Predicted': y_test_pred})
print(predic.round(3))
coef = pd.DataFrame({'Feature': training_1.iloc[:, :11].columns, 'Coefficient': lr1.coef_})
print(coef.round(3))
print()
print("Intercept:", round(lr1.intercept_,3))

     data from file   Predicted
0            280000  194207.932
1            520000  340719.587
2            150000  325416.849
3            180000  273672.748
4            300000  298369.367
..              ...         ...
745          330000  283138.706
746          450000  381114.180
747          180000  297490.123
748           90000  242061.854
749          360000  328403.653

[750 rows x 2 columns]
            Feature  Coefficient
0            Gender   -23183.330
1      10percentage      702.767
2      12percentage     1259.019
3       CollegeTier   -99570.608
4            Degree    18369.962
5        collegeGPA     1297.532
6   CollegeCityTier    -8836.727
7           English      141.760
8           Logical      145.742
9             Quant      114.643
10           Domain    34955.750

Intercept: 49248.09


### `Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra`

In [257]:
mae = calculate_mae(test, y_test_pred)
print("MAE: ", mae.round(3))

MAE:  105052.53


`Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân)`

 `Salary`= 49248.09 + (-23183.330)x`Gender `+ 702.767x`10percentage `+ 1259.019x`12percentage `</br> 
+(-99570.608)x`CollegeTier `+18369.962x`Degree `+1297.532x`collegeGPA `</br>
+(-8836.727)x`CollegeCityTier `+ 141.760x`English `</br>
+145.742x`Logical `+ 114.643x`Quant `+ 34955.750x`Domain` 

# Yêu cầu `1b`: Xây dựng mô hình sử dụng duy nhất 1 đặc trưng tính cách với các đặc trưng tính cách gồm `conscientiousness`, `agreeableness`, `extraversion`, `nueroticism`, `openess_to_experience`, tìm mô hình cho kết quả tốt nhất (1 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ đặc trưng

`Các bước làm gần như tương tự ở câu A, tuy nhiên bộ dữ liệu đặc trung lúc này là 5 tính cách kế cuối của bộ dữu liệu`

- Đọc dữ liệu huấn luyện từ tệp 'train.csv' và lưu vào biến train_2. Sau đó, dữ liệu huấn luyện được trộn ngẫu nhiên bằng phương thức sample(frac=1).
- Khởi tạo một đối tượng `LinearRegression `trong biến `reg`.
- Chọn 5 đặc trưng kế đặc trưng cuối cùng trong số 34 cột của dữ liệu huấn luyện và lưu vào danh sách `fea_title`.
- Với mỗi đặc trưng trong `fea_title`, thực hiện cross-validation bằng cách sử dụng hàm cross_val_score để tính toán điểm số sử dụng độ đo 'neg_mean_absolute_error' trên mô hình reg và các đặc trưng tương ứng. Kết quả cross-validation được lưu trong danh sách `scores`.
- Tính trung bình giá trị tuyệt đối của các điểm số cross-validation và lưu vào danh sách MAE.
- Tìm chỉ số của đặc trưng có giá trị MAE nhỏ nhất bằng hàm np.argmin(MAE) và lưu vào biến `best_index`.
- In ra màn hình đặc trưng có giá trị MAE nhỏ nhất và giá trị MAE tương ứng, số làm tròn đến 3 chữ số phần thập phân.

In [258]:
# Phần code cho yêu cầu 1b
# Tìm ra đặc trưng tốt nhất
# In ra các kết quả cross-validation như yêu cầu

training_2 = pd.read_csv('train.csv')
training_2 = training_2.sample(frac=1)
y_df = training_2.iloc[:, -1]

reg = LinearRegression()

fea_title = training_2.columns[-6:-1]  # Lấy 5 tính cách kế cuối của 34 cột
feature = [training_2[name].values.reshape(-1, 1) for name in fea_title]

cross = []
scores = []
MAE = []

for i in range(len(feature)):
    scores.append(cross_val_score(reg, feature[i], y_df, scoring="neg_mean_absolute_error", cv=5))


for i in range(len(scores)):
    MAE.append(np.absolute(scores[i]).mean())
    cross.append([fea_title[i], round(MAE[i], 3)])  # Làm tròn số thành 3 chữ số phần thập phân

test = pd.DataFrame(cross, columns=['Feature', 'Average MAE'])
print(test)

best_index = np.argmin(MAE)
print()
print("Minimum of average MAE is:", fea_title[best_index],'-',  round(MAE[best_index], 3))  

                 Feature  Average MAE
0      conscientiousness   124041.291
1          agreeableness   123518.689
2           extraversion   123694.831
3            nueroticism   123555.915
4  openess_to_experience   123770.125

Minimum of average MAE is: agreeableness - 123518.689


#### `Huấn luyện lại mô hình best_personality_feature_model với đặc trưng tốt nhất trên toàn bộ tập huấn luyện`

In [259]:
testing_2f = pd.read_csv('test.csv')

training_2f = pd.read_csv('train.csv')

X_k_test = testing_2f[fea_title[best_index]].values.reshape(-1, 1)
y_k_test = testing_2f.iloc[:, -1]

X_k_train = training_2f[fea_title[best_index]].values.reshape(-1, 1)
y_k_train = training_2f.iloc[:, -1]


reg2 = LinearRegression(fit_intercept=True).fit(X_k_train, y_k_train)

y_k_test_pred = reg2.predict(X_k_test)

k_predic = pd.DataFrame({'Data in file': y_k_test, 'Predicted': y_k_test_pred})
print(k_predic.round(3))

k_coef = pd.DataFrame({'Feature': fea_title[best_index], 'Coefficient': reg2.coef_})
print(k_coef.round(3))
print()
print("Intercept:", round(reg2.intercept_,3))


     Data in file   Predicted
0          280000  308026.917
1          520000  303086.416
2          150000  320378.169
3          180000  303127.587
4          300000  289945.000
..            ...         ...
745        330000  305764.104
746        450000  294438.956
747        180000  329492.760
748         90000  315437.668
749        360000  310497.167

[750 rows x 2 columns]
         Feature  Coefficient
0  agreeableness    15834.939

Intercept: 305037.28


#### `Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình best_personality_feature_model`

In [260]:
print("MAE: ", round(mean_absolute_error(y_k_test, y_k_test_pred),3))

MAE:  118153.163


`Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân)`

$$\text{Salary} = -16021.494 * nueroticism + 304647.553 $$

# Yêu cầu `1C`: Xây dựng mô hình sử dụng duy nhất 1 đặc trưng `English`, `Logical`, `Quant`, tìm mô hình cho kết quả tốt nhất (1 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ đặc trưng

`Các bước làm gần như tương tự ở câu B, tuy nhiên bộ dữ liệu đặc trung lúc này là các tính cách ở 3 vị trí từ 7 đến 9 của của bộ dữu liệu`

In [261]:
# Phần code cho yêu cầu 1c
# Tìm ra đặc trưng tốt nhất
# In ra các kết quả cross-validation như yêu cầu

training_2 = pd.read_csv('train.csv')
training_2 = training_2.sample(frac=1)
y_df = training_2.iloc[:, -1]
reg = LinearRegression()
fea_title = training_2.columns[7:10]  # Lấy từ 7 đến 9
feature = [training_2[name].values.reshape(-1, 1) for name in fea_title]
cross = []
scores = []
MAE = []

for i in range(len(feature)):
    scores.append(cross_val_score(reg, feature[i], y_df, scoring="neg_mean_absolute_error", cv=5))
    
# Calculate MAE and append to an array
for i in range(len(scores)):
    MAE.append(np.absolute(scores[i]).mean())
    cross.append([fea_title[i], round(MAE[i], 3)])  # Làm tròn số làm 3 chữ số phần thập phân
test = pd.DataFrame(cross, columns=['Feature', 'Average MAE'])
print(test)
print()
best_index = np.argmin(MAE)
print("Min average MAE is:", fea_title[best_index], "(", round(MAE[best_index], 3), ")")  # Làm tròn số làm 3 chữ số phần thập phân 


   Feature  Average MAE
0  English   120689.929
1  Logical   119885.620
2    Quant   117198.643

Min average MAE is: Quant ( 117198.643 )


`Huấn luyện lại mô hình best_personality_feature_model với đặc trưng tốt nhất trên toàn bộ tập huấn luyện`

In [262]:
training_2f = pd.read_csv('train.csv')
testing_2f = pd.read_csv('test.csv')

X_k_train = training_2f[fea_title[best_index]].values.reshape(-1, 1)
y_k_train = training_2f.iloc[:, -1]

X_k_test = testing_2f[fea_title[best_index]].values.reshape(-1, 1)
y_k_test = testing_2f.iloc[:, -1]

reg2 = LinearRegression(fit_intercept=True).fit(X_k_train, y_k_train)

y_k_test_pred = reg2.predict(X_k_test)

k_predic = pd.DataFrame({'Data in file': y_k_test, 'Predicted': y_k_test_pred})
print(k_predic.round(3))

k_coef = pd.DataFrame({'Feature': fea_title[best_index], 'Coefficient': reg2.coef_})
print(k_coef.round(3))
print()
print("Intercept:", round(reg2.intercept_,3))


     Data in file   Predicted
0          280000  197063.009
1          520000  359358.093
2          150000  337226.945
3          180000  270833.502
4          300000  302185.961
..            ...         ...
745        330000  302185.961
746        450000  326161.371
747        180000  245013.829
748         90000  322472.847
749        360000  311407.273

[750 rows x 2 columns]
  Feature  Coefficient
0   Quant      368.852

Intercept: 117759.729


`Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình best_personality_feature_model`

In [263]:
print("MAE: ", round(mean_absolute_error(y_k_test, y_k_test_pred),3))

MAE:  108814.06


Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

#### Salary = 368.852 * `Quant ` + 117759.729

# Yêu cầu `1D`: Sinh viên tự xây dựng mô hình, tìm mô hình cho kết quả tốt nhất (3 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ $m$ mô hình mà sinh viên thiết kế

## Tìm mô hình

#### Đầu tiên, ta sử dụng hàm `corr()` của pandas để tính hệ số tương quan giữa của từng cặp đôi một trong 23 đặc trưng để xem thử độ tương quan như thế nào.

In [264]:
# Phần code cho yêu cầu 1d
# Tìm ra mô hình tốt nhất (tự thiết kế bởi sinh viên)
# In ra các kết quả cross-validation như yêu cầu
# Trình bày các phần tìm ra mô hình
train_corr = pd.read_csv('train.csv')
correlations = train_corr.corr()
print(correlations)

                         Gender  10percentage  12percentage  CollegeTier   
Gender                 1.000000      0.165208      0.131372     0.028943  \
10percentage           0.165208      1.000000      0.644518    -0.135469   
12percentage           0.131372      0.644518      1.000000    -0.092920   
CollegeTier            0.028943     -0.135469     -0.092920     1.000000   
Degree                -0.007080     -0.255081     -0.227924    -0.014755   
collegeGPA             0.153008      0.311057      0.340745    -0.091280   
CollegeCityTier        0.044938      0.106144      0.120529    -0.093067   
English               -0.020830      0.335863      0.193790    -0.182399   
Logical               -0.000189      0.309735      0.240047    -0.189068   
Quant                 -0.104069      0.326948      0.316088    -0.240973   
Domain                 0.001947      0.079001      0.069002    -0.037476   
ComputerProgramming    0.021369      0.041760      0.076158    -0.047969   
ElectronicsA

 #### Tiếp theo, ta lấy ra hệ số tương quan giữa các đặc trưng với đặc trưng `Salary`.

In [265]:
correlations_2 = train_corr.corr()['Salary'].drop('Salary')
print(correlations_2)


Gender                  -0.036183
10percentage             0.155174
12percentage             0.149531
CollegeTier             -0.174824
Degree                  -0.017602
collegeGPA               0.122469
CollegeCityTier          0.004575
English                  0.169293
Logical                  0.188416
Quant                    0.205358
Domain                   0.122022
ComputerProgramming      0.125866
ElectronicsAndSemicon   -0.009292
ComputerScience         -0.095507
MechanicalEngg           0.028854
ElectricalEngg          -0.041217
TelecomEngg             -0.040415
CivilEngg                0.016150
conscientiousness       -0.057699
agreeableness            0.068623
extraversion            -0.002661
nueroticism             -0.073401
openess_to_experience   -0.007814
Name: Salary, dtype: float64


## Thử nghiệm, so sánh các mô hình

Để so sánh việc chuẩn hóa các đặc trưng ở phần code chính là tốt hay xấu, ta cần một kết quả riêng với mô hình .

In [266]:
train_3 = pd.read_csv('train.csv')
train_3 = train_3.sample(frac=1)

model = ['Use 23 features', 'Use 16 features', 'Use 10 features']
AMAE = []

### Sử dụng 5-Fold Cross Validation để test với tất cả 23 đặc trưng 

In [267]:
# 1st model
train_test_1 = train_3.copy()
feature = ['Gender', '10percentage', '12percentage', 'CollegeTier', 'Degree', 
           'collegeGPA', 'CollegeCityTier', 'English', 'Logical', 'Quant', 
           'Domain','ComputerProgramming','ElectronicsAndSemicon',
           'ComputerScience','MechanicalEngg','ElectricalEngg','TelecomEngg',
           'CivilEngg','conscientiousness','agreeableness',
           'extraversion','nueroticism','openess_to_experience']
X_test_1 = train_test_1[feature]
y_test_1 = train_test_1.iloc[:, -1]

rg_test_1 = LinearRegression()
scores_test_1 = cross_val_score(rg_test_1, X_test_1, y_test_1, cv=5, 
                                scoring='neg_mean_absolute_error')
print("Average MAE: ", round(np.absolute(scores_test_1).mean(),3))
AMAE.append(np.absolute(scores_test_1).mean())

Average MAE:  111856.239


### Sử dụng 5-Fold Cross Validation để test với 16 đặc trưng: `Gender`,`10percentage`, `12percentage` ,  `collegeGPA`, `English`, `Logical`, `Quant`, `Domain`, `ComputerProgramming`, `ComputerScience`, `MechanicalEngg`, `ElectricalEngg`, `TelecomEngg`, `conscientiousness`, `agreeableness`, `nueroticism` 

In [268]:
train_test_2 = train_3.copy()
feature = ['Gender','10percentage', '12percentage', 'collegeGPA',
           'English', 'Logical', 'Quant', 'Domain','ComputerProgramming',
           'ComputerScience','MechanicalEngg','ElectricalEngg','TelecomEngg',
           'conscientiousness','agreeableness','nueroticism']
X_test_2 = train_test_2[feature]
y_test_2 = train_test_2.iloc[:, -1]

rg_test_2 = LinearRegression()
scores_test_2 = cross_val_score(rg_test_2, X_test_2, y_test_2, cv=5, 
                                scoring='neg_mean_absolute_error')
print("Average MAE: ", round(np.absolute(scores_test_2).mean(),3))
AMAE.append(np.absolute(scores_test_2).mean())

Average MAE:  111191.237


### Sử dụng 5-Fold Cross Validation để test với 10 đặc trưng: `Gender`,`10percentage`, `12percentage` , `CollegeTier`, `collegeGPA`, `English`, `Logical`, `Quant`, `Domain`, `ComputerProgramming`, `ComputerScience` 

In [269]:
train_test_3 = train_3.copy()
feature = ['Gender','10percentage', '12percentage', 'CollegeTier',  
           'collegeGPA', 'English', 'Logical', 'Quant', 
           'Domain','ComputerProgramming','ComputerScience']
X_test_3 = train_test_3[feature]
y_test_3 = train_test_3.iloc[:, -1]

rg_test_3 = LinearRegression()
scores_test_3 = cross_val_score(rg_test_3, X_test_3, y_test_3, cv=5, 
                                scoring='neg_mean_absolute_error')
print("Average MAE: ", round(np.absolute(scores_test_3).mean(),3))
AMAE.append(np.absolute(scores_test_3).mean())


Average MAE:  111452.183


### So sánh các mô hình để chọn ra mô hình tối ưu nhất

In [270]:
Average_MAE = pd.DataFrame({'Model': model, 'Average MAE': AMAE})
print(round(Average_MAE,3))

             Model  Average MAE
0  Use 23 features   111856.239
1  Use 16 features   111191.237
2  Use 10 features   111452.183


### Chọn mô hình tối ưu nhất và huấn luyện lại mô hình my_best_model trên toàn bộ tập huấn luyện


In [271]:
train3 = pd.read_csv('train.csv')
test3 = pd.read_csv('test.csv')

feature = ['Gender','10percentage','12percentage', 'CollegeTier','collegeGPA', 
           'English', 'Logical', 'Quant', 'Domain','ComputerProgramming',
           'ComputerScience','MechanicalEngg','ElectricalEngg','TelecomEngg',
           'conscientiousness','agreeableness','nueroticism']
X_train_best_model = train3[feature].copy()
X_test_best_model = test3[feature].copy()

reg3 = LinearRegression(fit_intercept=True).fit(X_train_best_model, train.iloc[:, -1])

y_test_best_model_pred = reg3.predict(X_test_best_model)
predic_best_model = pd.DataFrame({'Actual': test3.iloc[:, -1], 
                                  'Predicted': y_test_best_model_pred})
print(predic_best_model)

coef_best_model = pd.DataFrame({'Coefficient': X_test_best_model.columns, 
                                'Predicted': reg3.coef_})
print(round(coef_best_model,3))
print()
print("Intercept:", round(reg3.intercept_,3))



     Actual      Predicted
0    280000  263567.962314
1    520000  356766.935756
2    150000  323743.600400
3    180000  217610.444880
4    300000  361556.869933
..      ...            ...
745  330000  318457.798212
746  450000  367203.243225
747  180000  248929.961391
748   90000  306346.947633
749  360000  332948.207394

[750 rows x 2 columns]
            Coefficient  Predicted
0                Gender -24909.837
1          10percentage    702.390
2          12percentage   1039.326
3           CollegeTier -97057.418
4            collegeGPA   1326.148
5               English    138.267
6               Logical    108.238
7                 Quant     91.541
8                Domain  25549.124
9   ComputerProgramming     90.459
10      ComputerScience   -172.478
11       MechanicalEngg     54.516
12       ElectricalEngg   -145.755
13          TelecomEngg    -89.379
14    conscientiousness -20081.590
15        agreeableness  14665.138
16          nueroticism -11465.937

Intercept: 99137.525


#### `Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình my_best_model`


In [272]:
print("MAE: ", round(mean_absolute_error(test3.iloc[:, -1], y_test_best_model_pred),3))

MAE:  103241.002


#### `Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân)`

Salary= 99137.525+ `Gender`* -24909.837 + `10percentage`* 702.390 + `12percentage `* 1039.326 </br> 
+`CollegeTier `* -97057.418 +`collegeGPA `* 1326.148 + `English`*138.267 +`Logical `* 108.238 </br>
+`Quant `* 91.541 +`Domain `* 25549.124 +`ComputerProgramming `* 90.459+ `ComputerScience `*  -172.478 </br>
+`MechanicalEngg `* 54.516 + `ElectricalEngg`* -145.755 +`TelecomEngg`* -89.379 + `conscientiousness`* -20081.590 </br>
+`agreeableness `*  14665.138 + `nueroticism `* -11465.937